tools/clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -2058,10 +2058,14 @@ QualType Sema::CheckTemplateIdType(TemplateName Name,
     return QualType();
 
   // HLSL Change Starts - check template values for HLSL object/matrix/vector signatures
-  if (getLangOpts().HLSL && Template->isImplicit() &&
-      hlsl::CheckTemplateArgumentListForHLSL(*this, Template, TemplateLoc,
-                                             TemplateArgs))
-    return QualType();
+  if (getLangOpts().HLSL && Template->isImplicit()) {
+    if (hlsl::CheckTemplateArgumentListForHLSL(*this, Template, TemplateLoc,
+                                               TemplateArgs))
+      return QualType();
+    if (hlsl::IsHLSLMatrixTemplate(*this, Template))
+      return hlsl::CheckHLSLMatrixTemplate(*this, Template, TemplateLoc,
```

Comment:
Should this be first, and the `hlsl::CheckTemplateArgumentListForHLSL` in an `else if` branch?  That way, all checks for the matrix template instantiation would be contained within `hlsl::CheckHLSLMatrixTemplate`, including any call to `hlsl::CheckTemplateArgumentListForHLSL` if necessary.

Reply:
Updated.